### PR TITLE
Stop taking locks when reading from unique indexes

### PR DESCRIPTION
**Pull request by systay**

### CODE REVIEW

**Comment by systay** [comment]
When merging this into 3.0, we should use this:
https://github.com/pontusmelke/neo4j/commit/33d399c19a8d80d5f80a6564422b3ea2e3195b92

It makes sure the cost planner doesn't take locks when doing read-only access.

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/UniqueIndexUsageAcceptanceTest.scala [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+class UniqueIndexUsageAcceptanceTest extends ExecutionEngineFunSuite with NewPlannerTestSupport {
+  test("should be able to use indexes") {
+    given()
+
+    // When
+    val result = executeWithAllPlanners("MATCH (n:Crew) WHERE n.name = 'Neo' RETURN n")
+
+    // Then
+    result.executionPlanDescription().toString should include("NodeUniqueIndexSeek")
```
Comment (pontusmelke) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/UniqueIndexUsageAcceptanceTest.scala:33]: ~~You can use `result should use("NodeUniqueIndexSeek")`~~

~~(likewise for all tests in this class)~~

Probably not in 2.2

File: community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v2_2/TransactionBoundQueryContext.scala [hunk 2]
```diff
@@ -131,10 +132,13 @@ final class TransactionBoundQueryContext(graph: GraphDatabaseAPI,
...
   }
 
-  def exactIndexSearch(index: IndexDescriptor, value: Any) =
+  def exactIndexSearch(index: IndexDescriptor, value: Any) = {
+    indexSearchMonitor.exactIndexSearch(index, value)
     mapToScala(statement.readOperations().nodesGetFromIndexLookup(index, value))(nodeOps.getById)
+  }
 
-  def exactUniqueIndexSearch(index: IndexDescriptor, value: Any): Option[Node] = {
+  def lockingIndexSearch(index: IndexDescriptor, value: Any): Option[Node] = {
```
Comment (lutovich) [inline community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v2_2/TransactionBoundQueryContext.scala:140]: Naming of these methods is a bit inconsistent. Maybe call this one `lockingExactIndexSearch`?
Or maybe we can name them `exactIndexSearchForRead` and `exactIndexSearchForWrite`.

**Comment by systay** [comment]
merged manually

**Pull request closed without merging**